crates/sui-json-rpc/src/coin_api.rs

```diff
@@ -19,118 +23,29 @@ use sui_open_rpc::Module;
 use sui_types::balance::Supply;
 use sui_types::base_types::{ObjectID, SuiAddress};
 use sui_types::coin::{CoinMetadata, TreasuryCap};
-use sui_types::effects::TransactionEffectsAPI;
-use sui_types::error::SuiError;
+use sui_types::effects::{TransactionEffects, TransactionEffectsAPI};
+use sui_types::error::{SuiError, SuiResult};
 use sui_types::gas_coin::GAS;
-use sui_types::object::Object;
+use sui_types::object::{Object, ObjectRead};
 use sui_types::parse_sui_struct_tag;
 
 use crate::api::{cap_page_limit, CoinReadApiServer, JsonRpcMetrics};
-use crate::error::{Error, SuiRpcInputError};
+use crate::error::{Error, SuiApiResult, SuiRpcInputError};
 use crate::{with_tracing, SuiRpcModule};
 
+#[cfg(test)]
+use mockall::automock;
+
 pub struct CoinReadApi {
-    state: Arc<AuthorityState>,
-    pub metrics: Arc<JsonRpcMetrics>,
+    internal: Arc<dyn CoinReadInternalTrait + Send + Sync>,
```

Comment:
Does this need to be `Arc`, and does it need to be a `dyn` trait?, or could we implement this statically, e.g.

```rust
pub struct CoinReadApi<Internal: CoinReadInternalTrait> {
    internal: Internal
}
```

Reply:
So in refactoring the code I had the option to either impl. these internal traits as a trait object or as a generic. I went with the former approach, which necessitates the use of `Arc<dyn`, namely for code verbosity, but if there's an inclination towards generics approach, then I also think the benefits of generics outweigh the slightly more code bloat

Reply:
There's also an option here to use `Box<dyn ...>` rather than `Arc<dyn ...>` -- `Arc` implies "shared ownership across multiple threads" vs `Box`'s "unique ownership", if there are benefits to using a trait object. (Put another way, when I see `Arc` it makes me wonder what else owns this internal object, and why its reference count needs to be protected by an atomic).

Reply:
Thanks Ashok, was not aware of `Box`. I think `Box` makes sense as we do not need to share the internal across multiple threads.